controllers/openstackdataplanenode_controller.go

```diff
@@ -266,3 +278,111 @@ func (r *OpenStackDataPlaneNodeReconciler) GenerateInventory(ctx context.Context
+	if instance.Spec.Node.Managed {
+		return strconv.FormatBool(instance.Spec.Node.Managed)
+	}
+	return strconv.FormatBool(instanceRole.Spec.NodeTemplate.Managed)
+}
+
+// GetAnsibleManagementNetwork returns the string value from the template unless it is set in the node
+func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleManagementNetwork(instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) string {
+	if instance.Spec.Node.ManagementNetwork != "" {
+		return instance.Spec.Node.ManagementNetwork
+	}
+	return instanceRole.Spec.NodeTemplate.ManagementNetwork
+}
+
+// GetAnsibleNetworkConfig returns a JSON string value from the template unless it is set in the node
+func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleNetworkConfig(instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) string {
+	if instance.Spec.Node.NetworkConfig != instanceRole.Spec.NodeTemplate.NetworkConfig {
+		return fmt.Sprintf("{template: %s}", instance.Spec.Node.NetworkConfig)
+
+	}
+	return fmt.Sprintf("{template: %s}", instanceRole.Spec.NodeTemplate.NetworkConfig.Template)
+}
+
+// GetAnsibleNetworks returns a JSON string mapping fixedIP and/or network name to their valules
+func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleNetworks(instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) string {
+	var network string // the resulting string containing each network
+	for _, netMap := range instance.Spec.Node.Networks {
+		if netMap.FixedIP != "" && netMap.Network == "" {
+			network += fmt.Sprintf("{%s: %s},", "fixedIP", netMap.FixedIP)
+		}
+		if netMap.FixedIP != "" && netMap.Network != "" {
+			network += fmt.Sprintf("{%s: %s, %s: %s},",
+				"fixedIP", netMap.FixedIP, "network", netMap.Network)
+		}
+	}
+	return fmt.Sprintf("[%s]", strings.TrimSuffix(network, ","))
+}
+
+// EnsureNodeOnRoleList will ensure that a node is on its role's dataPlaneNodes list
+func (r *OpenStackDataPlaneNodeReconciler) EnsureNodeOnRoleList(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) error {
+	// 1. When a node is created by a role it will already be on the list
+	// 2. When a node is directly created with a role name, it won't be on the list
+	// In the second case we have to add the node directly and set nodeFrom field
+
+	var err error
+	nodeOnList := false
+	for _, node := range instanceRole.Spec.DataPlaneNodes {
+		if node.NodeFrom == instance.Name {
+			nodeOnList = true
+			break
+		}
+		if node.HostName == instance.Spec.HostName {
+			nodeOnList = true
+			break
+		}
+		if node.AnsibleHost == instance.Spec.AnsibleHost {
+			nodeOnList = true
+			break
+		}
+	}
+	if !nodeOnList {
+		nodeSection := &dataplanev1beta1.OpenStackDataPlaneNodeSpec{}
+		_, err := controllerutil.CreateOrUpdate(ctx, r.Client, instanceRole, func() error {
```

Comment:
This needs to not fail as seen here  https://paste.opendev.org/show/bSM4Uy4xUQutVMFQH4lD/

Reply:
see:
- https://github.com/openstack-k8s-operators/osp-director-dev-tools/blob/master/ansible/patch_csv.yaml#L12-L15
- https://github.com/openstack-k8s-operators/osp-director-operator/blob/master/pkg/bindata_util/merge.go#L7-L21

Reply:
you probably need `MergeMetadataForUpdate` before calling the `Update`

Reply:
Maybe this is creating an infinite loop: reconcilenode --> reconcilerole which calls reconcilenode.

I'm going to remove `EnsureNodeOnRoleList` from the next update to this PR since it's not essential for the `dataPlaneNodes` list to reflect all of the nodes within the role.

Reply:
Thanks for the suggestions @fao89 maybe we can get `EnsureNodeOnRoleList` in a follow up PR.